shared/src/main/java/net/blay09/mods/waystones/block/entity/WarpPlateBlockEntity.java

```diff
@@ -198,19 +200,31 @@ private boolean isEntityOnWarpPlate(Entity entity) {
     }
 
     public void serverTick() {
-        if (isReadyForAttunement()) {
+        WarpPlateRecipe recipe = trySelectRecipe();
+        if (recipe != null) {
             attunementTicks++;
 
             if (attunementTicks >= getMaxAttunementTicks()) {
                 attunementTicks = 0;
-                ItemStack attunedShard = new ItemStack(ModItems.attunedShard);
+                ItemStack attunedShard = recipe.assemble(this, RegistryAccess.EMPTY);
                 WaystonesAPI.setBoundWaystone(attunedShard, getWaystone());
+                ItemStack centerStack = getItem(0);
+                if (centerStack.getCount() > 1) {
+                    centerStack = centerStack.copyWithCount(centerStack.getCount() - 1);
+                    if (!Minecraft.getInstance().player.getInventory().add(centerStack)) {
+                        Minecraft.getInstance().player.drop(centerStack, false);
+                    }
+                }
                 setItem(0, attunedShard);
                 for (int i = 1; i <= 4; i++) {
                     getItem(i).shrink(1);
                 }
-                completedFirstAttunement = true;
             }
+
+            if (recipe.getResultItem().getItem() == ModItems.attunedShard) {
```

Comment:
I think this check is not necessary because the first attunement will always happen first (since it handles locking the slots during the initial opening animation) and it doesn't harm "re"-setting it to true every time.